control/modelsimp.py

```diff
@@ -556,3 +557,216 @@ def markov(Y, U, m=None, transpose=False):
+            if not np.allclose(np.diff(data.time), dt):
+                raise ValueError("response time values must be equally "
+                                 "spaced.")
+        transpose = data.transpose
+        if data.transpose and not data.issiso:
+            Umat, Ymat = np.transpose(Umat), np.transpose(Ymat)
+        if len(args) == 2:
+            m = args[1]
+        elif len(args) > 2:
+            raise ControlArgument("too many positional arguments")
+    else:
+        if len(args) < 2:
+            raise ControlArgument("not enough input arguments")
+        Umat = np.array(args[1], ndmin=2)
+        Ymat = np.array(args[0], ndmin=2)
+        if dt is None:
+            dt = True
+        if transpose:
+            Umat, Ymat = np.transpose(Umat), np.transpose(Ymat)
+        if len(args) == 3:
+            m = args[2]
+        elif len(args) > 3:
+            raise ControlArgument("too many positional arguments")
+
+    # Make sure the number of time points match
+    if Umat.shape[1] != Ymat.shape[1]:
+        raise ControlDimension(
+            "Input and output data are of differnent lengths")
+    l = Umat.shape[1]
+
+    # If number of desired parameters was not given, set to size of input data
+    if m is None:
+        m = l
+
+    # Paper equation 8, Page 8
+    # There is a mistake in the paper, but it is right in the book
+    t = 0
+    if truncate:
+        t = m
+
+    # The okid in the paper estimates `m + 1` Markov parameters
+    # Change to `m` to match control.markov,
+    # TODO:What is the best way to match control.markov
+    # m = m - 1 
+
+    q = Ymat.shape[0] # number of outputs
+    p = Umat.shape[0] # number of inputs
+
+    # Make sure there is enough data to compute parameters
+    if m*p > (l-t):
+        warnings.warn("Not enough data for requested number of parameters")
+
+    # the algorithm - Construct a matrix of control virtual inputs to invert
+    #
+    # (q,l)   = (q,(p+q)*m+p) @ ((p+q)*m+p,l)
+    # YY.T    = Ybar @ VV.T
+    #
+    # This algorithm sets up the following problem and solves it for
+    # the observer Markov parameters
+    #
+    # (l,q)   = (l,(p+q)*m+p) @ ((p+q)*m+p,q) 
+    # YY      = VV @ Ybar.T
+    #
+    # truncated version t=m, do not use first m equation
+    #
+    # Note: This algorithm assumes C A^{j} B = 0
+    # for j > m-2.  See equation (3) in
+    #
+    #   J.-N. Juang, M. Phan, L. G. Horta, and R. W. Longman, Identification
+    #   of observer/Kalman filter Markov parameters - Theory and
+    #   experiments. Journal of Guidance Control and Dynamics, 16(2),
+    #   320-329, 2012. http://doi.org/10.2514/3.21006
+    #
+
+    Vmat = np.concatenate((Umat, Ymat),axis=0)
+
+    VVT = np.zeros(((p + q)*m + p, l))
+    
+    VVT[:p,:] = Umat
+    for i in range(m):
+        # Shift previous column down and keep zeros at the top
+        VVT[(p+q)*i+p:(p+q)*(i+1)+p, i+1:] = Vmat[:, :l-i-1]
+
+    YY = Ymat[:,t:].T
+    VV = VVT[:,t:].T
```

Comment:
Add spaces after commas in argument lists for consistency with standard Python style.  Here and below.